### PR TITLE
Merge/improved daily rest

### PR DESCRIPTION
**Pull request by hkgrov**

### CODE REVIEW

File: heuristic/repair_operators.py [hunk 1]
```diff
@@ -856,53 +857,79 @@ def mip_week_operator_2(  employees, shifts_in_week, competencies, time_periods_
                          state, destroy_set, week):
 
     days_in_week = days[7*week[0]:(week[0]+1)*7]
-    delta_calculate_negative_deviation_from_contracted_hours(state, employees, contracted_hours, weeks, time_periods_in_week, competencies, time_step)
-    updated_contracted_hours = min(sum(contracted_hours[e] for e in employees), sum(state.soft_vars["deviation_contracted_hours"][e,j] for j in weeks for e in employees))
+    days_in_week_x = [x_day for x_day in range(max(0, 7*week[0] - 1), min(7*(week[0] + 1) + 1, len(days)))]
+    #delta_calculate_negative_deviation_from_contracted_hours(state, employees, contracted_hours, weeks, time_periods_in_week, competencies, time_step)
+    
+    updated_time_periods = list({t for t_1, v_1 in shifts_in_week[week[0]] for t in t_covered_by_shift[t_1, v_1]})
+    updated_time_periods.sort()
```
Comment (ekallevik) [inline heuristic/repair_operators.py:864]: Why do you need sort?

File: heuristic/repair_operators.py [hunk 2]
```diff
@@ -919,17 +946,18 @@ def mip_week_operator_2(  employees, shifts_in_week, competencies, time_periods_
...
-        quicksum(delta_plus[c, t] + delta_minus[c, t]
-                for t in time_periods_in_week[c, week[0]])
+        quicksum(delta_plus[c, t] + delta_minus[c, t] for t in updated_time_periods)
         for c in competencies
     )
     ,GRB.MINIMIZE,)
 
     #model.setParam("MipFocus", 1)
-    model.setParam("TimeLimit", 10)
+    model.write("Restrictions.lp")
```
Comment (ekallevik) [inline heuristic/repair_operators.py:955]: Could you disable these writes?

File: heuristic/repair_operators.py [hunk 3]
```diff
@@ -919,17 +946,18 @@ def mip_week_operator_2(  employees, shifts_in_week, competencies, time_periods_
...
+    model.setParam("TimeLimit", 5)
     model.optimize()
+    model.write("answer.sol")
 
-    shifts = {(t,v): x[t,v].x for t,v in shifts_in_week[week[0]] if x[t,v].x > 0}
+    shifts = {(t,v): x[t,v].x for t,v in shifts_in_week[week[0]] if x[t,v].x > 0.5}
```
Comment (ekallevik) [inline heuristic/repair_operators.py:960]: What is the purpose here?

File: heuristic/repair_operators.py [hunk 4]
```diff
@@ -1064,24 +1092,38 @@ def mip_week_operator_3(    employees, shifts_in_week, competencies, time_period
                             state, destroy_set, week):
 
     days_in_week = days[7*week[0]:(week[0]+1)*7]
+    days_in_week_x = [x_day for x_day in range(max(0, 7*week[0] - 1), min(7*(week[0] + 1) + 1, len(days)))]
     delta_calculate_negative_deviation_from_contracted_hours(state, employees, contracted_hours, weeks, time_periods_in_week, competencies, time_step)
+    calculate_deviation_from_demand(state, competencies, t_covered_by_shift, employees_with_competencies, demand, destroy_set)
     updated_contracted_hours = {e: sum(state.soft_vars["deviation_contracted_hours"][e,j] for j in weeks) for e in employees}
     
+    updated_time_periods = list({t for t_1, v_1 in shifts_in_week[week[0]] for t in t_covered_by_shift[t_1, v_1]})
+    updated_time_periods.sort()
```
Comment (ekallevik) [inline heuristic/repair_operators.py:1101]: Necessary to sort?

File: heuristic/repair_operators.py (hunk 1 above)
Comment (hkgrov) [reply heuristic/repair_operators.py:864]: Most likely not needed. Can try without it

File: heuristic/repair_operators.py (hunk 2 above)
Comment (hkgrov) [reply heuristic/repair_operators.py:955]: Yes

File: heuristic/repair_operators.py (hunk 3 above)
Comment (hkgrov) [reply heuristic/repair_operators.py:960]: Gurobi does not always set = 1 for shifts that are used. This way we don't end up not using shifts that should be used

File: heuristic/repair_operators.py (hunk 4 above)
Comment (hkgrov) [reply heuristic/repair_operators.py:1101]: Most likely not

File: heuristic/repair_operators.py [hunk 5]
```diff
@@ -1193,18 +1277,17 @@ def mip_week_operator_3(    employees, shifts_in_week, competencies, time_period
...
         for e in employees)
-        - quicksum(delta_plus[c, t] + delta_minus[c, t] for c in competencies for t in time_periods_in_week[c, week[0]])
+        
+        - quicksum(delta_plus[c, t] + delta_minus[c, t] for c in competencies for t in updated_time_periods)
     , GRB.MAXIMIZE)
 
-
-    
-    #model.setParam("MIPFocus", 1)
+    model.write("restrictions.lp")
```
Comment (ekallevik) [inline heuristic/repair_operators.py:1288]: Remove

**Pull request merged**